### ISSUE TITLE
No longer works

### ISSUE THREAD

**Issue by elitistphoenix** [jensim/bitbucket_server_cli#37]
Hi,
Sorry, I can't give more information here as I can't find a debug option but the linux and windows excutables not longer work against on-prem Bitbucket v7.4.0

It logs in fine. Is able to find all the projects and repos fine. It just can't clone them. Error e.g:
tps/cookbook-01 failed git clone into .. Cause: Err: 'Cloning into 'cookbook-01'...
tps/cookbook-02 failed git clone into .. Cause: Err: 'Cloning into 'cookbook-02'...

I'm a full server admin. More than happy to try and debug the issue if you can tell me how.

**Comment by elitistphoenix** [comment]
I also can't find how it tries to clone the repos? Is it over the http link you set or does it try ssh?

**Comment by jensim** [comment]
I'm sorry it broke ☹️
Have you tried doing a git clone in the same terminal?
Since the cli is basically just doing that for you..?
If it prompts you for passwords or anything, that can be an indication what's wrong, since the CLI needs the ssh-agent to handle such auth

**Comment by jensim** [comment]
I've started working at another company since I made this tool, and no longer have access to a fat bit bucket server install to test this tool on 🤷 but I'm super open to PRs if you find bugs or potential improvements in the clone commands @elitistphoenix

**Comment by jensim** [comment]
> I also can't find how it tries to clone the repos? Is it over the http link you set or does it try ssh?

@elitistphoenix [Here is where the clone link is resolved from bitbucket - it depends on your preference](https://github.com/jensim/bitbucket_server_cli/blob/main/src/bitbucket/types.rs#L6)

[Here are the types, so that you can see the defaults, that should also be visible when you use the cli --help](https://sourcegraph.com/github.com/jensim/bitbucket_server_cli/-/blob/src/types.rs)

**Comment by elitistphoenix** [comment]
> CLI needs the ssh-agent

Ah so even if you initially specify http to for the server it'll try ssh? 

> Here are the types

My Rust coding isn't that great sorry. I'll try and get someone to help me make sense of it though.

> started working at another company

No problems. If I can help I'll make a PR.


Thanks for your help and time spent on this tool though.

**Comment by elitistphoenix** [comment]
CloneType::HTTP => "http"
So if it is http it tries http

**Comment by elitistphoenix** [comment]
> Have you tried doing a git clone in the same terminal?

Yes, a git clone http://server:port does work

(sorry for the tonne of replies)

**Comment by jensim** [comment]
> Yes, a git clone http://server:port does work

Does that prompt you for username and password?
If so you need to supply both of them to the CLI, or it will assume you dont need a password to clone

> (sorry for the tonne of replies)

No problem 😊

**Comment by jensim** [comment]
There's not much in the way of retry, or retry with something else built into this tool 😅 it's basically just makeup on a pig 🐷
It asks bit bucket for all the repos and delegates the cloning to your installed git command via rust Command..
If you set your concurrency to high it might make bit bucket choke. I experienced times when I sank the CI workers because I was holding all available git workers

**Comment by elitistphoenix** [comment]
> Does that prompt you for username and password?
Yes and it is able to login. I can see in the BB logs
Interesting some of these repos are "public" and it still fails to clone them too.

> makeup on a pig pig
That's what I'm aiming for right now! Dump 'em out!

I've tried locally(/lan) so it's not a timeout I don't believe.

**Comment by jensim** [comment]
Do you have multiple git installs @elitistphoenix ? 🤔

**Comment by elitistphoenix** [comment]
> multiple git installs

On the same server? No :)

**Comment by jensim** [comment]
It seems really odd to me that you dont get the same problems cloning manually. If you run it with the same user in the same shell.. Could you try using http clone with no username and no password, for just the public repos, and see what that gives you..?

**Comment by elitistphoenix** [comment]
Never used user/profile. New dir. Using HTTP it cloned one of the public repos perfectly.

**Comment by jensim** [comment]
> Never used user/profile. New dir. Using HTTP it cloned one of the public repos perfectly.

Sounds to me like you are not providing a password in a way that works for the HTTP clone.. 🤔 Could be a bug somewhere. I don't really have time to dig into it that deep though :( sorry that im not of that much help here

**Comment by elitistphoenix** [comment]
It does manage to login and get access to the  private repos though. Stuff that wouldn't be accessible without the password. I'll keep looking into it today.

**Comment by elitistphoenix** [comment]
I just found something. When I disabled SSH on the bitbucket server it said no repos to clone. It seems like it's trying to use ssh instead of http for the clone?

**Comment by jensim** [comment]
That's interesting...
Like I said, there's no fallback behavior built in..

If you inspect your own api returns, like they do in this gist:
https://gist.github.com/eeichinger/babbdf99d8fe21fef8634d56f29defba
(I do basically the same thing, but with rust instead of sh)
```
curl -s -u $USERNAME:$PASSWORD https://bitbucket.example.com/rest/api/1.0/projects/$PROJECTNAME/repos/ \
  | jq -r '.values[].links.clone[] | select(.name=="http") | .href'
```
What do you get?

Only thing I can think of is that your ssh links are also returned as `"name": "http"` in the `links>clone[]`-list 🤔 , otherwise the fallback would not make sense, as it does not exist in this project

**Comment by jensim** [comment]
I just raid your messages again, and I think I misunderstood you at first..
So there was no fallback behavior. But the CLI always went for ssh clone, and when you disabled it on the server side, it found no repos to clone at all.
I'm curious to find out what your `/rest/api/1.0/projects/$PROJECTNAME/repos/` call will return ^^

**Comment by elitistphoenix** [comment]
I've figured it out. I believe it was a mixture of bugs in bitbucket and maybe this tool too. Here are my findings for the next person:

I saw this is the highest of ssh debug logs "no mutual signature algorithm"
Doesn't seem right. My key is in there. Then I found this
https://confluence.atlassian.com/bitbucketserverkb/ssh-rsa-key-rejected-with-message-no-mutual-signature-algorithm-1026057701.html
then pointed to this
https://jira.atlassian.com/browse/BSERV-13013
grabbed my laptop which one has a ed25519 key on it. could clone via ssh
tl;dr they stuffed this version of ssh in bitbucket. instead of just disabling (rsa) sha1 they disabled sha256 and sha512 too which is what (mostly) all our old keys are
So, then tried the bitbucket clone tool on my laptop with only 1 ed25519 key and it worked.
Setup and ~/.ssh/config file on my local to force the ed25519 key for the server. Worked too.

However this I think does point to that this tool does try ssh over http even though http was specified at the start. Unfortunately I'm not a coder so I can't add a PR for this. Thanks for your awesome tool and help @jensim
